File: python/bob/ip/test/color.py
#!/usr/bin/env python
# vim: set fileencoding=utf-8 :
# Andre Anjos <andre.anjos@idiap.ch>
# Fri Mar 25 13:44:43 2011 +0100
#
# Copyright (C) 2011-2012 Idiap Research Institute, Martigny, Switzerland
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 3 of the License.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

"""Test color conversions available in bob
"""

import os, sys
import platform
import numpy
import unittest
import bob
import colorsys

class ColorTest(unittest.TestCase):
  """Various color conversion tests."""

  def test01_hsv(self):

    # This test verifies that color conversion is reversible for HSV <=> RGB
    # It also shows how you can convert single pixel representations from RGB
    # to HSV and vice-versa, testing the output of bob color converters
    # against python's.

    step = 0.02
    for r in numpy.arange(0, 1, step):
      for g in numpy.arange(0, 1, step):
        for b in numpy.arange(0, 1, step):
          # First test the correctness
          ht, st, vt = bob.ip.rgb_to_hsv_f(r, g, b)
          hp, sp, vp = colorsys.rgb_to_hsv(r, g, b)
          self.assertTrue(abs(ht - hp) < 1e-6)
          self.assertTrue(abs(st - sp) < 1e-6)
          self.assertTrue(abs(vt - vp) < 1e-6)
          # And that we can invert the result using bob
          r2, g2, b2 = bob.ip.hsv_to_rgb_f(ht, st, vt)
          self.assertTrue(abs(r2 - r) < 1e-6)
          self.assertTrue(abs(g2 - g) < 1e-6)
          self.assertTrue(abs(b2 - b) < 1e-6)

  def test02_hsl(self):

    # This test verifies that color conversion is reversible for HSL <=> RGB
    # It also shows how you can convert single pixel representations from RGB
    # to HSL and vice-versa, testing the output of bob color converters
    # against python's.

    step = 0.02
    for r in numpy.arange(0, 1, step):
      for g in numpy.arange(0, 1, step):
        for b in numpy.arange(0, 1, step):
          # First test the correctness
          ht, st, lt = bob.ip.rgb_to_hsl_f(r, g, b)
          hp, lp, sp = colorsys.rgb_to_hls(r, g, b)
          self.assertTrue(abs(ht - hp) < 1e-6)
          self.assertTrue(abs(st - sp) < 1e-6)
          self.assertTrue(abs(lt - lp) < 1e-6)
          # And that we can invert the result using bob
          r2, g2, b2 = bob.ip.hsl_to_rgb_f(ht, st, lt)
          self.assertTrue(abs(r2 - r) < 1e-6)
          self.assertTrue(abs(g2 - g) < 1e-6)
          self.assertTrue(abs(b2 - b) < 1e-6)

  def test03_yuv(self):

    # This test verifies that color conversion is reversible for YUV <=> RGB
    # It also shows how you can convert single pixel representations from RGB
    # to YUV and vice-versa.

    step = 0.02
    for r in numpy.arange(0, 1, step):
      for g in numpy.arange(0, 1, step):
        for b in numpy.arange(0, 1, step):
          # First test the correctness
          yt, ut, vt = bob.ip.rgb_to_yuv_f(r, g, b)
          # And that we can invert the result using bob
          r2, g2, b2 = bob.ip.yuv_to_rgb_f(yt, ut, vt)
          self.assertTrue(abs(r2 - r) < 1e-4)
          self.assertTrue(abs(g2 - g) < 1e-4)
          self.assertTrue(abs(b2 - b) < 1e-4)

  def test04_int_conversions(self):

    # You can also use integer based conversions in which case the ranges
    # should occupy the whole valid range for the type. We support unsigned
    # integers with 8 (uint8_t) or 16 bits (uint16_t). The ranges are 0 to 255
    # for 8-bit unsigned integers and 0 to 65535 for 16-bit unsigned ones. '0'
    # represents total black while the maximum value, total white. Internally,
    # bob converts the integers into float representations and calculate the
    # the conversions just like in tests 01 to 03 above. The last step is a
    # back conversion into integer scale. This procedure may lead differences
    # in the representations and the two-way conversion. 

    # Just test a subrange or the test will take too long

    # Expected errors
    #
    # OSX 10.6 |    HSV     |    HSL     |    YUV  
    # ---------+------------+------------+-------------
    # uint8_t  | (3) 1.18%  | (4) 1.57%  | (1) 0.39%  
    # uint16_t | (3) 0.005% | (4) 0.006% | (1) 0.006%

    mx = 4
    for r in range(0,5) + range(120,130) + range(253,256):
      for g in range(0,6) + range(125,135) + range(252,256):
        for b in range(0,7) + range(127,137) + range(252,256):
          ht, st, vt = bob.ip.rgb_to_hsv_u8(r, g, b)
          r2, g2, b2 = bob.ip.hsv_to_rgb_u8(ht, st, vt)
          #mx2 = max(abs(r2-r), abs(g2-g), abs(b2-b))
          #correct within a 2% margin 
          #if mx2 > mx and (mx2/255.) < 0.02: mx = mx2
          self.assertTrue(abs(r2 - r) <= mx)
          self.assertTrue(abs(g2 - g) <= mx)
          self.assertTrue(abs(b2 - b) <= mx)
    #print "uint8_t RGB/HSV/RGB error: %d (%.2f%%)" % (mx, 100*mx/255.) 

    mx = 5
    for r in range(0,5) + range(120,130) + range(253,256):
      for g in range(0,6) + range(125,135) + range(252,256):
        for b in range(0,7) + range(127,137) + range(252,256):
          ht, st, lt = bob.ip.rgb_to_hsl_u8(r, g, b)
          r2, g2, b2 = bob.ip.hsl_to_rgb_u8(ht, st, lt)
          #mx2 = max(abs(r2-r), abs(g2-g), abs(b2-b))
          #correct within a 2% margin 
          #if mx2 > mx and (mx2/255.) < 0.02: mx = mx2
          self.assertTrue(abs(r2 - r) <= mx)
          self.assertTrue(abs(g2 - g) <= mx)
          self.assertTrue(abs(b2 - b) <= mx)
    #print "uint8_t RGB/HSL/RGB error: %d (%.2f%%)" % (mx, 100*mx/255.) 

    mx = 2
    for r in range(0,5) + range(120,130) + range(253,256):
      for g in range(0,6) + range(125,135) + range(252,256):
        for b in range(0,7) + range(127,137) + range(252,256):
          yt, ut, vt = bob.ip.rgb_to_yuv_u8(r, g, b)
          r2, g2, b2 = bob.ip.yuv_to_rgb_u8(yt, ut, vt)
          #mx2 = max(abs(r2-r), abs(g2-g), abs(b2-b))
          #correct within a 2% margin 
          #if mx2 > mx and (mx2/255.) < 0.02: mx = mx2
          self.assertTrue(abs(r2 - r) <= mx)
          self.assertTrue(abs(g2 - g) <= mx)
          self.assertTrue(abs(b2 - b) <= mx)
    #print "uint8_t RGB/YCbCr/RGB error: %d (%.2f%%)" % (mx, 100*mx/255.)

    # Just test a subrange or the test will take too long
    mx = 3
    for r in range(0,5) + range(30000,30005) + range(65530,65536):
      for g in range(0,6) + range(30002,3007) + range(65525,65532):
        for b in range(0,7) + range(3003,3008) + range(65524,65531):
          ht, st, vt = bob.ip.rgb_to_hsv_u16(r, g, b)
          r2, g2, b2 = bob.ip.hsv_to_rgb_u16(ht, st, vt)
          #mx2 = max(abs(r2-r), abs(g2-g), abs(b2-b))
          #if mx2 > mx and (mx2/65535.) < 0.0001: mx = mx2
          self.assertTrue(abs(r2 - r) <= mx)
          self.assertTrue(abs(g2 - g) <= mx)
          self.assertTrue(abs(b2 - b) <= mx)
    #print "16-bit unsigned integer RGB/HSV/RGB error: %d (%.4f%%)" % (mx, 100*mx/65535.) 

    mx = 4
    for r in range(0,5) + range(30000,30005) + range(65530,65536):
      for g in range(0,6) + range(30002,3007) + range(65525,65532):
        for b in range(0,7) + range(3003,3008) + range(65524,65531):
          ht, st, lt = bob.ip.rgb_to_hsl_u16(r, g, b)
          r2, g2, b2 = bob.ip.hsl_to_rgb_u16(ht, st, lt)
          #mx2 = max(abs(r2-r), abs(g2-g), abs(b2-b))
          #if mx2 > mx and (mx2/65535.) < 0.0001: mx = mx2
          self.assertTrue(abs(r2 - r) <= mx)
          self.assertTrue(abs(g2 - g) <= mx)
          self.assertTrue(abs(b2 - b) <= mx)
    #print "16-bit unsigned integer RGB/HSL/RGB error: %d (%.4f%%)" % (mx, 100*mx/65535.) 

    mx = 4
    for r in range(0,10) + range(120,130) + range(250,256):
      for g in range(5,12) + range(125,135) + range(240,252):
        for b in range(7,15) + range(127,137) + range(235,251):
          yt, ut, vt = bob.ip.rgb_to_yuv_u16(r, g, b)
          r2, g2, b2 = bob.ip.yuv_to_rgb_u16(yt, ut, vt)
          #mx2 = max(abs(r2-r), abs(g2-g), abs(b2-b))
          #if mx2 > mx and (mx2/65535.) < 0.0001: mx = mx2
          self.assertTrue(abs(r2 - r) <= mx)
          self.assertTrue(abs(g2 - g) <= mx)
          self.assertTrue(abs(b2 - b) <= mx)
    #print "16-bit unsigned integer RGB/YCbCr/RGB error: %d (%.4f%%)" % (mx, 100*mx/65535.) 

  def test05_gray_halves(self):

    # tests some border-line cases for gray-scale conversion and makes sure
    # we don't have precision problems.

    if platform.architecture()[0] == '64bit': 
      # do a full test, require all values to be the same
      correct = bob.io.load('gray-u8-mids.hdf5')
      for k in range(correct.shape[0]):
        self.assertEqual(correct[k,3], 
            bob.ip.rgb_to_gray_u8(*[int(z) for z in correct[k,:3]]) 
            )
    else:
      # do a full test, require all values to be the same
      # use a special 32-bit file. About 1600 cases do not match the 64-bit
      # ones.
      correct = bob.io.load('gray-u8-mids-32bits.hdf5')
      for k in range(correct.shape[0]):
        self.assertEqual(correct[k,3], 
            bob.ip.rgb_to_gray_u8(*[int(z) for z in correct[k,:3]]) 
            )

# Instantiates our standard main module for unittests
main = bob.helper.unittest_main(ColorTest)

File: src/ip/cxx/color.cc
/**
 * @file ip/cxx/color.cc
 * @date Fri Mar 25 09:59:18 2011 +0100
 * @author Andre Anjos <andre.anjos@idiap.ch>
 *
 * @brief Implements many sorts of color transformations using standards
 *
 * Copyright (C) 2011-2012 Idiap Research Institute, Martigny, Switzerland
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3 of the License.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <cmath>
#include <limits>
#include <boost/format.hpp>
#include "bob/ip/color.h"

bob::ip::UnsupportedTypeForColorConversion::UnsupportedTypeForColorConversion(bob::core::array::ElementType eltype) throw(): m_eltype(eltype) {
}

bob::ip::UnsupportedTypeForColorConversion::~UnsupportedTypeForColorConversion() throw() {
}

const char* bob::ip::UnsupportedTypeForColorConversion::what() const throw() {
  try {
    boost::format message("Color conversion for type '%s' is not supported");
    message % bob::core::array::stringize(m_eltype);
    m_message = message.str();
    return m_message.c_str();
  } catch (...) {
    static const char* emergency = "bob::ip::UnsupportedTypeForColorConversion: cannot format, exception raised";
    return emergency;
  }
}

bob::ip::UnsupportedRowExtent::UnsupportedRowExtent(int expected, int got) throw(): 
  m_expected(expected),
  m_got(got)
{
}

bob::ip::UnsupportedRowExtent::~UnsupportedRowExtent() throw() {
}

const char* bob::ip::UnsupportedRowExtent::what() const throw() {
  try {
    boost::format message("Color conversion requires an array with %d rows, but I got %d instead");
    message % m_expected % m_got;
    m_message = message.str();
    return m_message.c_str();
  } catch (...) {
    static const char* emergency = "bob::ip::UnsupportedRowExtent: cannot format, exception raised";
    return emergency;
  }
}

/**
 * This method will scale and cast to integer a single double value, using the
 * standard library
 */
template <typename T>
static inline T scale (double value) {
  return static_cast<T>(rint(std::numeric_limits<T>::max()*value));
}

/**
 * This method will scale and cast to double a single integer value, using the
 * standard library
 */
template <typename T>
static inline double normalize (T value) {
  return static_cast<double>(value)/static_cast<double>(std::numeric_limits<T>::max());
}

/**
 * This method calculates the greatest value on the input 3-tuple
 */
template <typename T>
static T tmax (T c1, T c2, T c3) {
  if (c2 >= c3) {
    if (c1 >= c2) return c1;
    return c2;
  }
  if (c1 >= c3) return c1;
  return c3;
}

/**
 * This method calculates the lowest value on the input 3-tuple
 */
template <typename T>
static T tmin (T c1, T c2, T c3) {
  if (c2 <= c3) {
    if (c1 <= c2) return c1;
    return c2;
  }
  if (c1 <= c3) return c1;
  return c3;
}

/**
 * This method clamps the double value between 0 and 1
 */
static double clamp (double f) {
  return (f<0.)? 0. : (f>1.)? 1.: f;
}

template <> void bob::ip::rgb_to_hsv_one (uint8_t r, uint8_t g, uint8_t b,
    uint8_t& h, uint8_t& s, uint8_t& v) {
  double H, S, V;
  rgb_to_hsv_one(normalize(r), normalize(g), normalize(b), H, S, V);
  h = scale<uint8_t>(H); s = scale<uint8_t>(S); v = scale<uint8_t>(V);
}

template <> void bob::ip::rgb_to_hsv_one (uint16_t r, uint16_t g, uint16_t b,
    uint16_t& h, uint16_t& s, uint16_t& v) {
  double H, S, V;
  rgb_to_hsv_one(normalize(r), normalize(g), normalize(b), H, S, V);
  h = scale<uint16_t>(H); s = scale<uint16_t>(S); v = scale<uint16_t>(V);
}

template <> void bob::ip::rgb_to_hsv_one (double r, double g, double b,
    double& h, double& s, double& v) {
  v = tmax(r, g, b); //value
  
  //if the Value is 0, then we also set the other values to zero
  if (v == 0.0) { 
    h = s = v;
    return;
  }

  //computing the saturation
  double C = v - tmin(r, g, b);
  s = C / v;

  //if the Saturation value is zero, set Hue to zero and return
  if (s == 0.0) {
    h = s;
    return;
  }

  /**
   * 1) Hue' = (Green - Blue) / C, if Value == Red
   * 2) Hue' = 2 + (Blue - Red) / C, if Value == Green
   * 3) Hue' = 4 + (Red - Green) / C, if Value == Blue
   *
   * Note: Hue' varies between 0 and 6 using the above formulation, we multiply
   * C by 6 to make sure it varies between 0 and 1 (normalized).
   */
  C *= 6;
  if (v == r) {
    //When V == Red, we need to be careful because the Hue will wrap
    if (g >= b) h = clamp((g - b)/C); //first sextant
    else h = clamp(1 - ((b - g)/C)); //sextant 6
  }
  else if (v == g) h = clamp(1.0/3 + (b - r)/C); //sextants 2/3
  else h = clamp(2.0/3 + (r - g)/C); //sextants 4/5
}

template <> void bob::ip::hsv_to_rgb_one (uint8_t h, uint8_t s, uint8_t v,
    uint8_t& r, uint8_t& g, uint8_t& b) {
  double R, G, B;
  hsv_to_rgb_one(normalize(h), normalize(s), normalize(v), R, G, B);
  r = scale<uint8_t>(R); g = scale<uint8_t>(G); b = scale<uint8_t>(B);
}
  
template <> void bob::ip::hsv_to_rgb_one (uint16_t h, uint16_t s, uint16_t v,
    uint16_t& r, uint16_t& g, uint16_t& b) {
  double R, G, B;
  hsv_to_rgb_one(normalize(h), normalize(s), normalize(v), R, G, B);
  r = scale<uint16_t>(R); g = scale<uint16_t>(G); b = scale<uint16_t>(B);
}
  
template <> void bob::ip::hsv_to_rgb_one (double h, double s, double v,
    double& r, double& g, double& b) {
  
  if(s == 0) { // achromatic (gray)
    r = g = b = v;
    return;
  }

  const double Hp = 6*h;
  const uint8_t sextant = static_cast<uint8_t>(Hp);
  const double Hpmod2 = Hp - (2 * static_cast<uint8_t>(Hp/2)); //Hp%2
  double C = v * s;
  const double m = v - C;
  const double X = C * (1 - fabsf(Hpmod2 - 1)) + m;
  C += m;

  switch(sextant) {
    case 0: //Hue is between red and yellow (red + green)
      r = clamp(v);
      g = clamp(X);
      b = clamp(m);
      break;
    case 1: //Hue is between yellow (red + green) and green
      r = clamp(X);
      g = clamp(v);
      b = clamp(m);
      break;
    case 2: //Hue is between green and cyan (green + blue)
      r = clamp(m);
      g = clamp(v);
      b = clamp(X);
      break;
    case 3: //Hue is between cyan (green + blue) and blue
      r = clamp(m);
      g = clamp(X);
      b = clamp(v);
      break;
    case 4: //Hue is between blue and magenta (blue + red)
      r = clamp(X);
      g = clamp(m);
      b = clamp(v);
      break;
    default: //Hue is between magenta (blue + red) and red
      r = clamp(v);
      g = clamp(m);
      b = clamp(X);
      break;
  }
}

template <> void bob::ip::rgb_to_hsl_one (uint8_t r, uint8_t g, uint8_t b,
    uint8_t& h, uint8_t& s, uint8_t& l) {
  double H, S, L;
  rgb_to_hsl_one(normalize(r), normalize(g), normalize(b), H, S, L);
  h = scale<uint8_t>(H); s = scale<uint8_t>(S); l = scale<uint8_t>(L);
}

template <> void bob::ip::rgb_to_hsl_one (uint16_t r, uint16_t g, uint16_t b,
    uint16_t& h, uint16_t& s, uint16_t& l) {
  double H, S, L;
  rgb_to_hsl_one(normalize(r), normalize(g), normalize(b), H, S, L);
  h = scale<uint16_t>(H); s = scale<uint16_t>(S); l = scale<uint16_t>(L);
}

template <> void bob::ip::rgb_to_hsl_one (double r, double g, double b,
    double& h, double& s, double& l) {
  //lightness calculation: L = (M + m)/2
  const double M = tmax(r, g, b);
  const double m = tmin(r, g, b);
  l = 0.5 * (M+m);
  
  //if the lightness is 0, then we also set the other values to zero
  if (l == 0) { 
    h = s = l;
    return;
  }

  //computing the saturation based on the lightness:
  //S = C / (1 - |2*L -1|)
  double C = M - m; //chroma
  double delta = 1-fabsf(2*l - 1);
  if (delta == 0)
    s = 0;
  else
    s = clamp(C / delta); 

  //if the Saturation value is zero, set Hue to zero and return
  if (s == 0) {
    h = s;
    return;
  }

  /**
   * 1) Hue' = (Green - Blue) / C, if Value == Red
   * 2) Hue' = 2 + (Blue - Red) / C, if Value == Green
   * 3) Hue' = 4 + (Red - Green) / C, if Value == Blue
   *
   * Note: Hue' varies between 0 and 6 using the above formulation, we multiply
   * C by 6 to make sure it varies between 0 and 1 (normalized).
   */
  C *= 6;
  if (M == r) {
    //When V == Red, we need to be careful because the Hue will wrap
    if (g >= b) h = clamp((g - b)/C); //first sextant
    else h = clamp(1 - ((b - g)/C)); //sextant 6
  }
  else if (M == g) h = clamp(1.0/3 + (b - r)/C); //sextants 2/3
  else h = clamp(2.0/3 + (r - g)/C); //sextants 4/5
}

template <> void bob::ip::hsl_to_rgb_one (uint8_t h, uint8_t s, uint8_t l,
    uint8_t& r, uint8_t& g, uint8_t& b) {
  double R, G, B;
  hsl_to_rgb_one(normalize(h), normalize(s), normalize(l), R, G, B);
  r = scale<uint8_t>(R); g = scale<uint8_t>(G); b = scale<uint8_t>(B);
}
  
template <> void bob::ip::hsl_to_rgb_one (uint16_t h, uint16_t s, uint16_t l,
    uint16_t& r, uint16_t& g, uint16_t& b) {
  double R, G, B;
  hsl_to_rgb_one(normalize(h), normalize(s), normalize(l), R, G, B);
  r = scale<uint16_t>(R); g = scale<uint16_t>(G); b = scale<uint16_t>(B);
}
  
template <> void bob::ip::hsl_to_rgb_one (double h, double s, double l,
    double& r, double& g, double& b) {
  
  double C = s*(1-fabsf(2*l - 1)); //Chroma [0,1]
  const double v = (2*l + C)/2; //Value [0,1]
  
  if(v == 0.) { // achromatic (gray)
    r = g = b = v; //Value
    return;
  }

  const double Hp = 6*h;
  const uint8_t sextant = static_cast<uint8_t>(Hp);
  const double Hpmod2 = Hp - (2 * static_cast<uint8_t>(Hp/2)); //Hp%2
  const double m = l - C/2;
  const double X = C * (1 - fabsf(Hpmod2 - 1)) + m;
  C += m;

  switch(sextant) {
    case 0: //Hue is between red and yellow (red + green)
      r = clamp(v);
      g = clamp(X);
      b = clamp(m);
      break;
    case 1: //Hue is between yellow (red + green) and green
      r = clamp(X);
      g = clamp(v);
      b = clamp(m);
      break;
    case 2: //Hue is between green and cyan (green + blue)
      r = clamp(m);
      g = clamp(v);
      b = clamp(X);
      break;
    case 3: //Hue is between cyan (green + blue) and blue
      r = clamp(m);
      g = clamp(X);
      b = clamp(v);
      break;
    case 4: //Hue is between blue and magenta (blue + red)
      r = clamp(X);
      g = clamp(m);
      b = clamp(v);
      break;
    default: //Hue is between magenta (blue + red) and red
      r = clamp(v);
      g = clamp(m);
      b = clamp(X);
      break;
  }
}

template <> void bob::ip::rgb_to_yuv_one (uint8_t r, uint8_t g, uint8_t b,
    uint8_t& y, uint8_t& u, uint8_t& v) {
  double Y, U, V;
  rgb_to_yuv_one(normalize(r), normalize(g), normalize(b), Y, U, V);
  y = scale<uint8_t>(Y); u = scale<uint8_t>(U); v = scale<uint8_t>(V);
}

template <> void bob::ip::rgb_to_yuv_one (uint16_t r, uint16_t g, uint16_t b,
    uint16_t& y, uint16_t& u, uint16_t& v) {
  double Y, U, V;
  rgb_to_yuv_one(normalize(r), normalize(g), normalize(b), Y, U, V);
  y = scale<uint16_t>(Y); u = scale<uint16_t>(U); v = scale<uint16_t>(V);
}

/**
 * Using the JPEG YUV conversion scheme
 */
template <> void bob::ip::rgb_to_yuv_one (double r, double g, double b,
    double& y, double& u, double& v) {
  bob::ip::rgb_to_gray_one(r, g, b, y); //Y'
  u = clamp(0.5 - 0.168736*r - 0.331264*g + 0.5*b); //Cb [0, 1]
  v = clamp(0.5 + 0.5*r - 0.418688*g - 0.081312*b); //Cr [0, 1]
}

template <> void bob::ip::yuv_to_rgb_one (uint8_t y, uint8_t u, uint8_t v,
    uint8_t& r, uint8_t& g, uint8_t& b) {
  double R, G, B;
  yuv_to_rgb_one(normalize(y), normalize(u), normalize(v), R, G, B);
  r = scale<uint8_t>(R); g = scale<uint8_t>(G); b = scale<uint8_t>(B);
}
  
template <> void bob::ip::yuv_to_rgb_one (uint16_t y, uint16_t u, uint16_t v,
    uint16_t& r, uint16_t& g, uint16_t& b) {
  double R, G, B;
  yuv_to_rgb_one(normalize(y), normalize(u), normalize(v), R, G, B);
  r = scale<uint16_t>(R); g = scale<uint16_t>(G); b = scale<uint16_t>(B);
}
  
/**
 * We are doing the inverse of the rgb_to_yuv_one() method above
 */
template <> void bob::ip::yuv_to_rgb_one (double y, double u, double v,
    double& r, double& g, double& b) {
  r = clamp(y + 1.40199959*(v-0.5));
  b = clamp(y + 1.772000066*(u-0.5));
  g = clamp(y - 0.344135678*(u-0.5) - 0.714136156*(v-0.5));
}

template <> void bob::ip::rgb_to_gray_one (uint8_t r, uint8_t g, uint8_t b,
    uint8_t& y) {
  double Y;
  rgb_to_gray_one(normalize(r), normalize(g), normalize(b), Y);
  y = scale<uint8_t>(Y);
}

template <> void bob::ip::rgb_to_gray_one (uint16_t r, uint16_t g, uint16_t b,
    uint16_t& y) {
  double Y;
  rgb_to_gray_one(normalize(r), normalize(g), normalize(b), Y);
  y = scale<uint16_t>(Y);
}

/**
 * Y = 0.299R+0.587G+0.114B
 */
template <> void bob::ip::rgb_to_gray_one (double r, double g, double b, 
    double& gray) {
  gray = clamp(0.299*r + 0.587*g + 0.114*b);
}
